### PR TITLE
Add api definitions

### PR DESCRIPTION
**Pull request by brahmaroutu**
This PR adds hack/codegen.sh to generate the API controller artifacts for all the API's defined under apis/objectstorage.k8s.io/v1alpha1. All CRD definitions along with listers informers and clientset are also checked-in for this package to be consumable as is.

@wlan0 @xing-yang Please review

### CODE REVIEW

**Review by wlan0 — CHANGES_REQUESTED** [review]

File: apis/objectstorage.k8s.io/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type BucketAccessRequestList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BucketAccessRequest `json:"items"`
+}
+
+type ObjectReference struct {
```
Comment (wlan0) [inline apis/objectstorage.k8s.io/v1alpha1/types.go:278]: Why are we reinventing this?

**Review by wlan0 — CHANGES_REQUESTED** [review]

File: apis/objectstorage.k8s.io/v1alpha1/types.go (hunk 1 above)
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/types.go:278]: We are defining a smaller footprint than the core.ObjectReference. Currently it is used as common struct to represent BucketRequest in BucketSpec and PolicyActionsConfigMap in BucketAccessClass. These two references can as well be v1alpha1.BucketRequest and core.ConfigMap. 
Please suggest.

File: apis/objectstorage.k8s.io/v1alpha1/s3_types.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// +k8s:deepcopy-gen=false
+
+package v1alpha1
+
+type S3SignatureVersion string
+
+const (
+	S3SignatureVersionV2 = "s3v2"
+	S3SignatureVersionV4 = "s3v4"
```
[the remark this answers is not in the document]
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/s3_types.go:9]: Fixed

File: apis/objectstorage.k8s.io/v1alpha1/s3_types.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// +k8s:deepcopy-gen=false
+
+package v1alpha1
+
+type S3SignatureVersion string
+
+const (
+	S3SignatureVersionV2 = "s3v2"
```
[the remark this answers is not in the document]
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/s3_types.go:8]: ok, changes it to uppercase as suggested.

File: apis/objectstorage.k8s.io/v1alpha1/protocol_types.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	ProtocolNameS3    ProtocolName = "s3"
+	ProtocolNameAzure ProtocolName = "azureBlob"
+	ProtocolNameGCS   ProtocolName = "gcs"
+)
+
+type RequestedProtocol struct {
+	// +kubebuilder:validation:Enum:={s3,azureBlob,gcs}
+	Name ProtocolName `json:"name"`
+	// +optional
+	Version string `json:"version"`
```
[the remark this answers is not in the document]
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/protocol_types.go]: Fixed

File: apis/objectstorage.k8s.io/v1alpha1/azure_types.go [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// +k8s:deepcopy-gen=false
```
[the remark this answers is not in the document]
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/azure_types.go:17]: fixed

File: apis/objectstorage.k8s.io/v1alpha1/doc.go [hunk 6]
```diff
@@ -0,0 +1,7 @@
+// +k8s:deepcopy-gen=package,register
```
[the remark this answers is not in the document]
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/doc.go:17]: fixed

File: apis/objectstorage.k8s.io/v1alpha1/openapi_generated.go [hunk 7]
```diff
@@ -0,0 +1,1236 @@
...
+package v1alpha1
+
+import (
+	spec "github.com/go-openapi/spec"
+	common "k8s.io/kube-openapi/pkg/common"
+)
+
+func GetOpenAPIDefinitions(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition {
+	return map[string]common.OpenAPIDefinition{
+		"github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1.AnonymousAccessMode":       schema_container_object_storage_interface_api_apis_objectstoragek8sio_v1alpha1_AnonymousAccessMode(ref),
```
[the remark this answers is not in the document]
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/openapi_generated.go:31]: I don't see how it is point to old repo, can you please explain.

File: apis/objectstorage.k8s.io/v1alpha1/types.go [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+type BucketAccessSpec struct {
+	// +optional
+	BucketInstanceName string `json:"bucketInstanceName,omitempty"`
+	// +optional
+	BucketAccessRequest string `json:"bucketAccessRequest,omitempty"`
+	// +optional
+	ServiceAccount string `json:"serviceAccount,omitempty"`
+	// +optional
+	MintedSecretName string `json:"mintedSecretName,omitempty"`
+
```
Comment (wlan0) [inline apis/objectstorage.k8s.io/v1alpha1/types.go:207]: nit: any reason for the space here?

File: apis/objectstorage.k8s.io/v1alpha1/types.go [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+
+type BucketAccess struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
```
Comment (wlan0) [inline apis/objectstorage.k8s.io/v1alpha1/types.go:237]: empty line

File: apis/objectstorage.k8s.io/v1alpha1/types.go [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+	BucketInstanceName string `json:"bucketInstanceName,omitempty"`
+	// +optional
+	BucketAccessRequest string `json:"bucketAccessRequest,omitempty"`
+	// +optional
+	ServiceAccount string `json:"serviceAccount,omitempty"`
+	// +optional
+	MintedSecretName string `json:"mintedSecretName,omitempty"`
+
+	PolicyActionsConfigMapData string `json:"policyActionsConfigMapData,omitempty"`
+
```
Comment (wlan0) [inline apis/objectstorage.k8s.io/v1alpha1/types.go:209]: empty line

File: apis/objectstorage.k8s.io/v1alpha1/types.go [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+	// +optional
+	Principal string `json:"principal,omitempty"`
+
```
Comment (wlan0) [inline apis/objectstorage.k8s.io/v1alpha1/types.go:212]: empty line

File: apis/objectstorage.k8s.io/v1alpha1/types.go (hunk 1 above)
Comment (wlan0) [reply apis/objectstorage.k8s.io/v1alpha1/types.go:294]: Makes sense. Can we call it BucketReference

**Comment by wlan0** [comment]
Mostly just nits, and one concern with ObjectReference. Otherwise, LGTM.

File: apis/objectstorage.k8s.io/v1alpha1/types.go (hunk 9 above)
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/types.go]: fixed

File: apis/objectstorage.k8s.io/v1alpha1/types.go (hunk 11 above)
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/types.go]: fixed

File: apis/objectstorage.k8s.io/v1alpha1/types.go (hunk 10 above)
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/types.go]: fixed

File: apis/objectstorage.k8s.io/v1alpha1/types.go (hunk 8 above)
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/types.go]: removed all empty lines

File: apis/objectstorage.k8s.io/v1alpha1/types.go (hunk 1 above)
Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/types.go]: fixed, added as per the suggestion

**Comment by brahmaroutu** [comment]
@wlan0 @rrati  
Can you please review

**Review by rrati — APPROVED** [review]
lgtm

**Review by wlan0 — APPROVED** [review]

File: apis/objectstorage.k8s.io/v1alpha1/azure_types.go [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=false
+
+package v1alpha1
```
Comment (jeffvance) [inline apis/objectstorage.k8s.io/v1alpha1/azure_types.go:19]: Do we need protocol-specific types now? I may be a bit behind, but I thought the `Bucket` instance was no longer going to expose protocol-specific fields, other than, perhaps, common values for all protocols.

Comment (brahmaroutu) [reply apis/objectstorage.k8s.io/v1alpha1/azure_types.go:19]: @jeffvance This PR is inline with the KEP that is merged. We will iterate it when the new proposal is merged.

**Pull request merged**